File: erlang/src/copilot_types.erl
%%%-------------------------------------------------------------------
%%% @doc Type specifications and record definitions for the Copilot SDK.
%%%
%%% Provides records, type specs, and constants used throughout
%%% the SDK. All public types are exported for use in client code.
%%% @end
%%%-------------------------------------------------------------------
-module(copilot_types).

%% API exports
-export([
    connection_state/1,
    session_event_type/1,
    tool_result/2,
    tool_result/3,
    permission_result/1,
    permission_result/2,
    user_input_response/2,
    image_options/1,
    elicitation_result/1,
    elicitation_result/2,
    tool_defer/1,
    system_message_section/1,
    github_attachment/1,
    hook_type/1,
    session_limits/1,
    memory_configuration/1,
    client_options_params/1,
    agent_definition/1
]).

%% ---------------------------------------------------------------------------
%% Records
%% ---------------------------------------------------------------------------

-record(session_fs_config, {
    initial_cwd        :: binary(),
    session_state_path :: binary(),
    conventions        :: binary()
}).

-record(command_definition, {
    name        :: binary(),
    description :: binary() | undefined,
    handler     :: fun()
}).

-record(client_options, {
    cli_path         :: binary() | undefined,
    cli_url          :: binary() | undefined,
    cwd              :: binary() | undefined,
    log_level        :: binary() | undefined,
    github_token     :: binary() | undefined,
    use_logged_in    :: boolean() | undefined,
    auto_start       :: boolean() | undefined,
    auto_restart     :: boolean() | undefined,
    session_idle_timeout_seconds :: non_neg_integer() | undefined,
    session_fs       :: #session_fs_config{} | undefined,
    copilot_home     :: binary() | undefined,
    tcp_connection_token :: binary() | undefined,
    request_handler       :: fun() | undefined,
    bearer_token_provider :: fun() | undefined,
    builtin_plugin_directories :: [binary()] | undefined,
    in_process            :: boolean() | undefined
}).

-record(session_config, {
    model                :: binary() | undefined,
    reasoning_effort     :: binary() | undefined,
    tools                :: [map()] | undefined,
    system_message       :: map() | undefined,
    available_tools      :: [binary()] | undefined,
    excluded_tools       :: [binary()] | undefined,
    working_directory    :: binary() | undefined,
    streaming            :: boolean() | undefined,
    on_permission_request :: fun() | undefined,
    on_user_input_request :: fun() | undefined,
    hooks                :: map() | undefined,
    provider             :: map() | undefined,
    mcp_servers          :: map() | undefined,
    infinite_sessions    :: map() | undefined,
    skill_directories    :: [binary()] | undefined,
    disabled_skills      :: [binary()] | undefined,
    include_sub_agent_streaming_events :: boolean() | undefined,
    model_capabilities   :: map() | undefined,
    enable_config_discovery :: boolean() | undefined,
    github_token         :: binary() | undefined,
    commands             :: [#command_definition{}] | undefined,
    on_elicitation_request :: fun() | undefined,
    instruction_directories :: [binary()] | undefined,
    %% Upstream-sync session options (parity with @github/copilot-sdk)
    enable_citations         :: boolean() | undefined,
    excluded_builtin_agents  :: [binary()] | undefined,
    session_limits           :: map() | undefined,
    memory_config            :: map() | undefined,
    otlp_protocol            :: binary() | undefined,
    enable_web_socket_responses :: boolean() | undefined,
    exp_assignments          :: map() | undefined,
    rewind_enabled           :: boolean() | undefined,
    additional_directories   :: [binary()] | undefined,
    disabled_mcp_servers     :: [binary()] | undefined,
    github_mcp_tool_config   :: map() | undefined,
    canvas_provider          :: map() | undefined,
    custom_agents_local_only :: boolean() | undefined,
    tool_search              :: map() | undefined,
    experimental_mode        :: boolean() | undefined,
    content_exclusion        :: boolean() | undefined,
    on_mcp_auth_request      :: fun() | undefined
}).

-record(command_context, {
    session_id    :: binary(),
    command       :: binary(),
    command_name  :: binary(),
    args          :: binary()
}).

-record(elicitation_context, {
    session_id          :: binary(),
    message             :: binary(),
    requested_schema    :: map() | undefined,
    mode                :: binary() | undefined,
    elicitation_source  :: binary() | undefined,
    url                 :: binary() | undefined
}).

-record(elicitation_result, {
    action  :: binary(),
    content :: map() | undefined
}).

-record(session_event, {
    type :: binary(),
    data :: map() | undefined,
    id :: binary() | undefined,
    timestamp :: binary() | undefined,
    parent_id :: binary() | undefined,
    agent_id :: binary() | undefined,
    ephemeral :: boolean() | undefined
}).

-record(tool_definition, {
    name        :: binary(),
    description :: binary(),
    parameters  :: map() | undefined,
    handler     :: fun()
}).

-record(tool_result, {
    text_result_for_llm :: binary(),
    result_type         :: success | failure,
    error               :: binary() | undefined,
    tool_telemetry      :: map() | undefined
}).

-record(tool_invocation, {
    session_id   :: binary(),
    tool_call_id :: binary(),
    tool_name    :: binary(),
    arguments    :: map()
}).

-record(permission_request_result, {
    kind :: approved | denied
}).

-record(user_input_response, {
    answer       :: binary(),
    was_freeform :: boolean()
}).

-record(image_options, {
    size    :: binary() | undefined,
    quality :: binary() | undefined,
    style   :: binary() | undefined
}).

%% Export record types for external use
-export_type([
    client_options/0,
    session_config/0,
    session_event/0,
    tool_definition/0,
    tool_result/0,
    tool_invocation/0,
    permission_request_result/0,
    user_input_response/0,
    image_options/0,
    connection_state/0,
    session_event_type/0,
    session_fs_config/0,
    command_definition/0,
    command_context/0,
    elicitation_context/0,
    elicitation_result/0
]).

-type client_options()           :: #client_options{}.
-type session_config()           :: #session_config{}.
-type session_event()            :: #session_event{}.
-type tool_definition()          :: #tool_definition{}.
-type tool_result()              :: #tool_result{}.
-type tool_invocation()          :: #tool_invocation{}.
-type permission_request_result() :: #permission_request_result{}.
-type user_input_response()      :: #user_input_response{}.
-type image_options()            :: #image_options{}.
-type session_fs_config()        :: #session_fs_config{}.
-type command_definition()       :: #command_definition{}.
-type command_context()          :: #command_context{}.
-type elicitation_context()      :: #elicitation_context{}.
-type elicitation_result()       :: #elicitation_result{}.
-type connection_state()         :: disconnected | connecting | connected | error.
-type session_event_type()       :: binary().

%% ---------------------------------------------------------------------------
%% Connection state constants
%% ---------------------------------------------------------------------------

-spec connection_state(atom()) -> binary().
connection_state(disconnected) -> <<"disconnected">>;
connection_state(connecting)   -> <<"connecting">>;
connection_state(connected)    -> <<"connected">>;
connection_state(error)        -> <<"error">>.

%% ---------------------------------------------------------------------------
%% Session event type constants
%% ---------------------------------------------------------------------------

-spec session_event_type(atom()) -> binary().
session_event_type(abort)                         -> <<"abort">>;
session_event_type(assistant_intent)              -> <<"assistant.intent">>;
session_event_type(assistant_message)             -> <<"assistant.message">>;
session_event_type(assistant_message_delta)       -> <<"assistant.message_delta">>;
session_event_type(assistant_reasoning)           -> <<"assistant.reasoning">>;
session_event_type(assistant_reasoning_delta)     -> <<"assistant.reasoning_delta">>;
session_event_type(assistant_turn_end)            -> <<"assistant.turn_end">>;
session_event_type(assistant_turn_start)          -> <<"assistant.turn_start">>;
session_event_type(assistant_usage)               -> <<"assistant.usage">>;
session_event_type(hook_end)                      -> <<"hook.end">>;
session_event_type(hook_start)                    -> <<"hook.start">>;
session_event_type(pending_messages_modified)     -> <<"pending_messages.modified">>;
session_event_type(session_compaction_complete)   -> <<"session.compaction_complete">>;
session_event_type(session_compaction_start)      -> <<"session.compaction_start">>;
session_event_type(session_error)                 -> <<"session.error">>;
session_event_type(session_handoff)               -> <<"session.handoff">>;
session_event_type(session_idle)                  -> <<"session.idle">>;
session_event_type(session_info)                  -> <<"session.info">>;
session_event_type(session_model_change)          -> <<"session.model_change">>;
session_event_type(session_resume)                -> <<"session.resume">>;
session_event_type(session_shutdown)              -> <<"session.shutdown">>;
session_event_type(session_snapshot_rewind)       -> <<"session.snapshot_rewind">>;
session_event_type(session_start)                 -> <<"session.start">>;
session_event_type(session_truncation)            -> <<"session.truncation">>;
session_event_type(session_usage_info)            -> <<"session.usage_info">>;
session_event_type(skill_invoked)                 -> <<"skill.invoked">>;
session_event_type(subagent_completed)            -> <<"subagent.completed">>;
session_event_type(subagent_failed)               -> <<"subagent.failed">>;
session_event_type(subagent_selected)             -> <<"subagent.selected">>;
session_event_type(subagent_started)              -> <<"subagent.started">>;
session_event_type(system_message)                -> <<"system.message">>;
session_event_type(tool_execution_complete)       -> <<"tool.execution_complete">>;
session_event_type(tool_execution_partial_result) -> <<"tool.execution_partial_result">>;
session_event_type(tool_execution_progress)       -> <<"tool.execution_progress">>;
session_event_type(tool_execution_start)          -> <<"tool.execution_start">>;
session_event_type(tool_user_requested)           -> <<"tool.user_requested">>;
session_event_type(user_message)                  -> <<"user.message">>;
session_event_type(command_execute)               -> <<"command.execute">>;
session_event_type(elicitation_requested)         -> <<"elicitation.requested">>;
session_event_type(capabilities_changed)          -> <<"capabilities.changed">>.

%% ---------------------------------------------------------------------------
%% Constructor helpers
%% ---------------------------------------------------------------------------

-spec tool_result(binary(), success | failure) -> map().
tool_result(Text, Type) ->
    tool_result(Text, Type, undefined).

-spec tool_result(binary(), success | failure, binary() | undefined) -> map().
tool_result(Text, Type, Error) ->
    Base = #{
        <<"textResultForLlm">> => Text,
        <<"resultType">> => atom_to_binary(Type, utf8)
    },
    case Error of
        undefined -> Base;
        _         -> Base#{<<"error">> => Error}
    end.

-spec permission_result(approved | denied) -> map().
permission_result(Kind) ->
    #{<<"kind">> => atom_to_binary(Kind, utf8)}.

%% @doc Permission reply including an opaque decision context surfaced to the CLI.
-spec permission_result(approved | denied, map() | undefined) -> map().
permission_result(Kind, undefined) ->
    permission_result(Kind);
permission_result(Kind, DecisionContext) ->
    (permission_result(Kind))#{<<"decisionContext">> => DecisionContext}.

-spec user_input_response(binary(), boolean()) -> map().
user_input_response(Answer, WasFreeform) ->
    #{
        <<"answer">>      => Answer,
        <<"wasFreeform">> => WasFreeform
    }.

-spec image_options(map()) -> map().
image_options(Opts) ->
    maps:filter(fun(_K, V) -> V =/= undefined end, #{
        <<"size">>    => maps:get(size, Opts, undefined),
        <<"quality">> => maps:get(quality, Opts, undefined),
        <<"style">>   => maps:get(style, Opts, undefined)
    }).

-spec elicitation_result(binary()) -> map().
elicitation_result(Action) ->
    elicitation_result(Action, undefined).

-spec elicitation_result(binary(), map() | undefined) -> map().
elicitation_result(Action, Content) ->
    Base = #{<<"action">> => Action},
    case Content of
        undefined -> Base;
        _         -> Base#{<<"content">> => Content}
    end.

%% ---------------------------------------------------------------------------
%% Upstream-sync feature constants & helpers (parity with @github/copilot-sdk)
%% ---------------------------------------------------------------------------

%% @doc Tool "defer" loading policy: lazy via search (auto) or eager pre-load (never).
-spec tool_defer(auto | never) -> binary().
tool_defer(auto)  -> <<"auto">>;
tool_defer(never) -> <<"never">>.

%% @doc System-message section identifiers. The preamble section targets only the
%% identity preamble; preserve protects an addressable section from removal.
-spec system_message_section(atom()) -> binary().
system_message_section(preamble)          -> <<"preamble">>;
system_message_section(identity)          -> <<"identity">>;
system_message_section(tool_instructions) -> <<"tool_instructions">>;
system_message_section(preserve)          -> <<"preserve">>.

%% @doc GitHub-anchored attachment variants.
-spec github_attachment(atom()) -> binary().
github_attachment(github_commit)          -> <<"GitHubCommit">>;
github_attachment(github_release)         -> <<"GitHubRelease">>;
github_attachment(github_actions_job)     -> <<"GitHubActionsJob">>;
github_attachment(github_repository)      -> <<"GitHubRepository">>;
github_attachment(github_file_diff)       -> <<"GitHubFileDiff">>;
github_attachment(github_tree_comparison) -> <<"GitHubTreeComparison">>;
github_attachment(github_url)             -> <<"GitHubUrl">>;
github_attachment(github_file)            -> <<"GitHubFile">>;
github_attachment(github_snippet)         -> <<"GitHubSnippet">>.

%% @doc Hook name constants (include post_tool_use and pre_mcp_tool_call).
-spec hook_type(atom()) -> binary().
hook_type(pre_tool_use)          -> <<"preToolUse">>;
hook_type(post_tool_use)         -> <<"postToolUse">>;
hook_type(pre_mcp_tool_call)     -> <<"preMcpToolCall">>;
hook_type(user_prompt_submitted) -> <<"userPromptSubmitted">>;
hook_type(user_prompt_transformed) -> <<"userPromptTransformed">>;
hook_type(session_start)         -> <<"sessionStart">>;
hook_type(session_end)           -> <<"sessionEnd">>.

%% @doc Build a sessionLimits wire map from a per-session AI-credit budget.
-spec session_limits(map()) -> map().
session_limits(Opts) ->
    maps:filter(fun(_K, V) -> V =/= undefined end, #{
        <<"maxAiCredits">> => maps:get(max_ai_credits, Opts, undefined)
    }).

%% @doc Build a memory configuration wire map (opt-in persistent session memory).
-spec memory_configuration(map()) -> map().
memory_configuration(Opts) ->
    maps:filter(fun(_K, V) -> V =/= undefined end, #{
        <<"enabled">> => maps:get(enabled, Opts, undefined)
    }).

%% @doc Build a wire map of transport-level client options (builtin plugin
%% directories and in-process FFI transport), parity with @github/copilot-sdk.
-spec client_options_params(map()) -> map().
client_options_params(Opts) ->
    maps:filter(fun(_K, V) -> V =/= undefined end, #{
        <<"builtinPluginDirectories">> => maps:get(builtin_plugin_directories, Opts, undefined),
        <<"inProcess">>                => maps:get(in_process, Opts, undefined)
    }).

%% @doc Build a custom-agent/factory definition wire map, including the optional
%% input argument JSON schema (argsSchema).
-spec agent_definition(map()) -> map().
agent_definition(Opts) ->
    maps:filter(fun(_K, V) -> V =/= undefined end, #{
        <<"name">>        => maps:get(name, Opts, undefined),
        <<"description">> => maps:get(description, Opts, undefined),
        <<"argsSchema">>  => maps:get(args_schema, Opts, undefined)
    }).

File: fortran/src/copilot_types.f90
! -------------------------------------------------------------------------------------
!  Copyright (c) Microsoft Corporation. All rights reserved.
! -------------------------------------------------------------------------------------

!> Derived types for the Copilot SDK.
!! Defines options, configuration, metadata, events, tools, and permissions.
module copilot_types
  implicit none
  private

  ! Re-export all public types
  public :: copilot_client_options, session_config, session_metadata
  public :: session_event, send_options, send_and_wait_result
  public :: copilot_tool, tool_result, tool_invocation
  public :: permission_request, permission_result
  public :: elicitation_request, elicitation_result
  public :: exit_plan_mode_request_t, exit_plan_mode_response_t
  public :: session_fs_config, session_fs_entry
  public :: copilot_connection_state
  public :: mcp_server_config, command_definition, image_options
  public :: COPILOT_STATE_DISCONNECTED, COPILOT_STATE_CONNECTING
  public :: COPILOT_STATE_CONNECTED, COPILOT_STATE_ERROR
  public :: tool_callback_interface
  ! --- Upstream-sync feature surface (parity with @github/copilot-sdk) ---
  public :: session_limits_config, memory_configuration, provider_token_args
  public :: bearer_token_provider_interface, copilot_request_handler_interface
  public :: mcp_auth_handler_interface, post_tool_use_hook_interface
  public :: pre_mcp_tool_call_hook_interface
  public :: TOOL_DEFER_AUTO, TOOL_DEFER_NEVER
  public :: SECTION_PREAMBLE, SECTION_IDENTITY, SECTION_TOOL_INSTRUCTIONS, SECTION_PRESERVE
  public :: GITHUB_COMMIT, GITHUB_RELEASE, GITHUB_ACTIONS_JOB, GITHUB_REPOSITORY
  public :: GITHUB_FILE_DIFF, GITHUB_TREE_COMPARISON, GITHUB_URL, GITHUB_FILE, GITHUB_SNIPPET

  ! --------------------------------------------------------------------------
  ! Connection state enum
  ! --------------------------------------------------------------------------
  integer, parameter :: COPILOT_STATE_DISCONNECTED = 0
  integer, parameter :: COPILOT_STATE_CONNECTING   = 1
  integer, parameter :: COPILOT_STATE_CONNECTED    = 2
  integer, parameter :: COPILOT_STATE_ERROR        = 3

  ! --------------------------------------------------------------------------
  ! Upstream-sync string constants
  ! --------------------------------------------------------------------------
  !> Tool "defer" loading policy (ToolDefer): eager pre-load or lazy search.
  character(len=*), parameter :: TOOL_DEFER_AUTO  = 'auto'   ! lazy: discovered on demand
  character(len=*), parameter :: TOOL_DEFER_NEVER = 'never'  ! eager: always pre-loaded

  !> System-message section identifiers ('preamble' targets the identity preamble;
  !! 'preserve' shields a section from a group-level remove).
  character(len=*), parameter :: SECTION_PREAMBLE          = 'preamble'
  character(len=*), parameter :: SECTION_IDENTITY          = 'identity'
  character(len=*), parameter :: SECTION_TOOL_INSTRUCTIONS = 'tool_instructions'
  character(len=*), parameter :: SECTION_PRESERVE          = 'preserve'

  !> GitHub-anchored attachment variants.
  character(len=*), parameter :: GITHUB_COMMIT          = 'GitHubCommit'
  character(len=*), parameter :: GITHUB_RELEASE         = 'GitHubRelease'
  character(len=*), parameter :: GITHUB_ACTIONS_JOB     = 'GitHubActionsJob'
  character(len=*), parameter :: GITHUB_REPOSITORY      = 'GitHubRepository'
  character(len=*), parameter :: GITHUB_FILE_DIFF       = 'GitHubFileDiff'
  character(len=*), parameter :: GITHUB_TREE_COMPARISON = 'GitHubTreeComparison'
  character(len=*), parameter :: GITHUB_URL             = 'GitHubUrl'
  character(len=*), parameter :: GITHUB_FILE            = 'GitHubFile'
  character(len=*), parameter :: GITHUB_SNIPPET         = 'GitHubSnippet'

  type :: copilot_connection_state
    integer :: value = COPILOT_STATE_DISCONNECTED
  end type copilot_connection_state

  ! --------------------------------------------------------------------------
  ! Client options
  ! --------------------------------------------------------------------------
  type :: copilot_client_options
    character(len=:), allocatable :: cli_path
    character(len=:), allocatable :: cli_url
    character(len=:), allocatable :: cwd
    character(len=:), allocatable :: log_level
    character(len=:), allocatable :: github_token
    logical :: auto_start       = .true.
    logical :: auto_restart     = .false.
    logical :: use_logged_in_user = .false.
    logical :: use_stdio        = .true.
    integer :: session_idle_timeout_seconds = 0
    type(session_fs_config), allocatable :: session_fs
    character(len=:), allocatable :: copilot_home
    character(len=:), allocatable :: tcp_connection_token
    !> Enable remote session support (Mission Control); adds the --remote CLI flag.
    logical :: remote = .false.
    !> HTTP request handler to intercept/mutate outbound LLM inference requests
    !! (CopilotRequestHandler).
    procedure(copilot_request_handler_interface), pointer, nopass :: request_handler => null()
    !> BYOK bearer token provider (per-session token minting).
    procedure(bearer_token_provider_interface), pointer, nopass :: bearer_token_provider => null()
  contains
    procedure :: set_defaults => client_options_set_defaults
  end type copilot_client_options

  ! --------------------------------------------------------------------------
  ! Session filesystem entry
  ! --------------------------------------------------------------------------
  type :: session_fs_entry
    character(len=:), allocatable :: path
    character(len=:), allocatable :: content
    logical :: is_directory = .false.
  end type session_fs_entry

  ! --------------------------------------------------------------------------
  ! Session filesystem configuration
  ! --------------------------------------------------------------------------
  type :: session_fs_config
    character(len=:), allocatable :: initial_cwd
    character(len=:), allocatable :: session_state_path
    character(len=:), allocatable :: conventions
  end type session_fs_config

  ! --------------------------------------------------------------------------
  ! MCP server configuration
  ! --------------------------------------------------------------------------
  type :: mcp_server_config
    character(len=:), allocatable :: command
    character(len=:), allocatable :: server_type
    character(len=:), allocatable :: url
    integer :: timeout = 0
  end type mcp_server_config

  ! --------------------------------------------------------------------------
  ! Command definition
  ! --------------------------------------------------------------------------
  type :: command_definition
    character(len=:), allocatable :: name
    character(len=:), allocatable :: description
  end type command_definition

  ! --------------------------------------------------------------------------
  ! Image options
  ! --------------------------------------------------------------------------
  type :: image_options
    character(len=:), allocatable :: size
    character(len=:), allocatable :: quality
    character(len=:), allocatable :: style
  end type image_options

  ! --------------------------------------------------------------------------
  ! Per-session AI-credit spending limits (SessionLimitsConfig)
  ! --------------------------------------------------------------------------
  type :: session_limits_config
    integer :: max_ai_credits = 0
  end type session_limits_config

  ! --------------------------------------------------------------------------
  ! Persistent session memory configuration (MemoryConfiguration)
  ! --------------------------------------------------------------------------
  type :: memory_configuration
    logical :: enabled = .false.
  end type memory_configuration

  ! --------------------------------------------------------------------------
  ! Arguments passed to a BYOK bearer_token_provider (per-session scoping)
  ! --------------------------------------------------------------------------
  type :: provider_token_args
    character(len=:), allocatable :: session_id
  end type provider_token_args

  ! --------------------------------------------------------------------------
  ! Session configuration
  ! --------------------------------------------------------------------------
  type :: session_config
    character(len=:), allocatable :: model
    character(len=:), allocatable :: system_prompt
    character(len=:), allocatable :: github_token
    character(len=:), allocatable :: workspace_path
    character(len=:), allocatable :: session_id
    type(copilot_tool), allocatable :: tools(:)
    type(session_fs_config) :: fs
    logical :: streaming = .true.
    integer :: session_idle_timeout_seconds = 0
    character(len=:), allocatable :: skill_directories(:)
    character(len=:), allocatable :: disabled_skills(:)
    character(len=:), allocatable :: excluded_tools(:)
    logical :: include_sub_agent_streaming_events = .false.
    logical :: enable_config_discovery = .false.
    character(len=:), allocatable :: model_capabilities_json
    character(len=:), allocatable :: request_headers_json
    character(len=:), allocatable :: mcp_servers_json
    type(command_definition), allocatable :: commands(:)
    type(image_options) :: img_options
    character(len=:), allocatable :: response_format
    character(len=:), allocatable :: instruction_directories(:)
    ! --- Upstream-sync session options (parity with @github/copilot-sdk) ---
    logical :: enable_citations = .false.
    character(len=:), allocatable :: excluded_builtin_agents(:)
    type(session_limits_config) :: session_limits
    type(memory_configuration) :: memory
    character(len=:), allocatable :: otlp_protocol
    logical :: enable_web_socket_responses = .false.
    character(len=:), allocatable :: exp_assignments_json
    !> Post-tool-use hook (inspect/override a tool result).
    procedure(post_tool_use_hook_interface), pointer, nopass :: on_post_tool_use => null()
    !> Pre-MCP-tool-call hook (gate an MCP server tool call).
    procedure(pre_mcp_tool_call_hook_interface), pointer, nopass :: on_pre_mcp_tool_call => null()
    !> MCP OAuth host token handler; when set, session.create signals mcpAuthHandler=true.
    procedure(mcp_auth_handler_interface), pointer, nopass :: on_mcp_auth_request => null()
    ! --- Newly-synced session options (parity with @github/copilot-sdk) ---
    !> Enable session telemetry (wire: enableSessionTelemetry).
    logical :: enable_session_telemetry = .false.
    !> Enable conversational rewind (restore an earlier turn) (wire: rewindEnabled).
    logical :: rewind_enabled = .false.
    !> Extra workspace directories to expose (wire: additionalDirectories).
    character(len=:), allocatable :: additional_directories(:)
    !> MCP servers to disable for this session (wire: disabledMcpServers).
    character(len=:), allocatable :: disabled_mcp_servers(:)
    !> GitHub MCP tool config as a JSON string (wire: githubMcpToolConfig).
    character(len=:), allocatable :: github_mcp_tool_config_json
    !> Canvas provider config as a JSON string (wire: canvasProvider).
    character(len=:), allocatable :: canvas_provider_json
    !> Restrict custom agents to locally-defined ones only (wire: customAgentsLocalOnly).
    logical :: custom_agents_local_only = .false.
    !> Permission-reply decision context as a JSON string (wire: decisionContext).
    character(len=:), allocatable :: decision_context_json
    !> Built-in plugin directories to load (wire: builtinPluginDirectories).
    character(len=:), allocatable :: builtin_plugin_directories(:)
    !> Agent-factory authoring arguments schema as a JSON string (wire: argsSchema).
    character(len=:), allocatable :: args_schema_json
    !> Reasoning effort control, e.g. 'low'/'medium'/'high' (wire: reasoningEffort).
    character(len=:), allocatable :: reasoning_effort
    !> Tool-search configuration as a JSON string (wire: toolSearch).
    character(len=:), allocatable :: tool_search_json
    !> Use the in-process FFI transport instead of a spawned CLI (wire: inProcess).
    logical :: in_process = .false.
    !> Enable experimental mode features (wire: experimentalMode).
    logical :: experimental_mode = .false.
    !> Enable content exclusion policy enforcement (wire: contentExclusion).
    logical :: content_exclusion = .false.
    !> User-prompt-transformed hook; when set, session.create signals userPromptTransformed=true.
    procedure(post_tool_use_hook_interface), pointer, nopass :: on_user_prompt_transformed => null()
  end type session_config

  ! --------------------------------------------------------------------------
  ! Session metadata
  ! --------------------------------------------------------------------------
  type :: session_metadata
    character(len=:), allocatable :: session_id
    character(len=:), allocatable :: model
    character(len=:), allocatable :: status
    character(len=:), allocatable :: created_at
    integer :: message_count = 0
  end type session_metadata

  ! --------------------------------------------------------------------------
  ! Session event
  ! --------------------------------------------------------------------------
  type :: session_event
    character(len=:), allocatable :: event_type
    character(len=:), allocatable :: id
    character(len=:), allocatable :: timestamp
    character(len=:), allocatable :: parent_id
    character(len=:), allocatable :: agent_id
    logical :: ephemeral = .false.
    character(len=:), allocatable :: session_id
    character(len=:), allocatable :: data_json
  end type session_event

  ! --------------------------------------------------------------------------
  ! Send options
  ! --------------------------------------------------------------------------
  type :: send_options
    character(len=:), allocatable :: message
    character(len=:), allocatable :: command
    character(len=:), allocatable :: model
    !> Per-send agent mode override (wire key: agentMode)
    character(len=:), allocatable :: agent_mode
    !> Prompt text shown to the user instead of the real prompt (wire key: displayPrompt)
    character(len=:), allocatable :: display_prompt
    !> Custom HTTP headers for this send as a JSON object string (wire key: requestHeaders)
    character(len=:), allocatable :: request_headers_json
  end type send_options

  ! --------------------------------------------------------------------------
  ! Send-and-wait result
  ! --------------------------------------------------------------------------
  type :: send_and_wait_result
    character(len=:), allocatable :: message
    character(len=:), allocatable :: error_message
    logical :: is_error = .false.
  end type send_and_wait_result

  ! --------------------------------------------------------------------------
  ! Tool invocation context
  ! --------------------------------------------------------------------------
  type :: tool_invocation
    character(len=:), allocatable :: invocation_id
    character(len=:), allocatable :: session_id
    character(len=:), allocatable :: arguments_json
  end type tool_invocation

  ! --------------------------------------------------------------------------
  ! Tool result
  ! --------------------------------------------------------------------------
  type :: tool_result
    character(len=:), allocatable :: content
    character(len=:), allocatable :: error_text
    logical :: is_error = .false.
  end type tool_result

  ! --------------------------------------------------------------------------
  ! Tool definition
  ! --------------------------------------------------------------------------
  type :: copilot_tool
    character(len=:), allocatable :: name
    character(len=:), allocatable :: description
    character(len=:), allocatable :: input_schema_json
    !> Defer loading policy: 'auto' (lazy) or 'never' (eager). See TOOL_DEFER_* constants.
    character(len=:), allocatable :: defer
    procedure(tool_callback_interface), pointer, nopass :: callback => null()
  end type copilot_tool

  ! --------------------------------------------------------------------------
  ! Permission request / result
  ! --------------------------------------------------------------------------
  type :: permission_request
    character(len=:), allocatable :: kind
    character(len=:), allocatable :: description
    character(len=:), allocatable :: resource
    character(len=:), allocatable :: session_id
  end type permission_request

  type :: permission_result
    logical :: granted = .false.
    character(len=:), allocatable :: reason
  end type permission_result

  ! --------------------------------------------------------------------------
  ! Elicitation request / result
  ! --------------------------------------------------------------------------
  type :: elicitation_request
    character(len=:), allocatable :: session_id
    character(len=:), allocatable :: message
    character(len=:), allocatable :: mode
    character(len=:), allocatable :: schema_json
  end type elicitation_request

  type :: elicitation_result
    character(len=:), allocatable :: action
    character(len=:), allocatable :: content_json
  end type elicitation_result

  ! --------------------------------------------------------------------------
  ! Exit plan mode request / response
  ! --------------------------------------------------------------------------
  type :: exit_plan_mode_request_t
    character(len=:), allocatable :: session_id
    character(len=:), allocatable :: summary
    character(len=:), allocatable :: plan_content
    character(len=:), allocatable :: recommended_action
  end type exit_plan_mode_request_t

  type :: exit_plan_mode_response_t
    logical :: approved = .false.
    character(len=:), allocatable :: selected_action
    character(len=:), allocatable :: feedback
  end type exit_plan_mode_response_t

  ! --------------------------------------------------------------------------
  ! Abstract interfaces for callbacks
  ! --------------------------------------------------------------------------
  abstract interface
    function tool_callback_interface(args, invocation) result(res)
      import :: tool_invocation, tool_result
      character(len=*), intent(in) :: args
      type(tool_invocation), intent(in) :: invocation
      type(tool_result) :: res
    end function tool_callback_interface

    !> BYOK bearer token provider: mint a per-session bearer token.
    function bearer_token_provider_interface(args) result(token)
      import :: provider_token_args
      type(provider_token_args), intent(in) :: args
      character(len=:), allocatable :: token
    end function bearer_token_provider_interface

    !> HTTP request handler: intercept/mutate an outbound LLM inference request
    !! (JSON string in, possibly-modified JSON string out).
    function copilot_request_handler_interface(request_json) result(modified_json)
      character(len=*), intent(in) :: request_json
      character(len=:), allocatable :: modified_json
    end function copilot_request_handler_interface

    !> MCP OAuth host token handler: return an OAuth token for an MCP server request.
    function mcp_auth_handler_interface(session_id, server_url) result(token)
      character(len=*), intent(in) :: session_id, server_url
      character(len=:), allocatable :: token
    end function mcp_auth_handler_interface

    !> Post-tool-use hook: inspect a tool result and optionally return an override.
    function post_tool_use_hook_interface(tool_name, tool_result_json, session_id) result(override)
      character(len=*), intent(in) :: tool_name, tool_result_json, session_id
      character(len=:), allocatable :: override
    end function post_tool_use_hook_interface

    !> Pre-MCP-tool-call hook: gate an MCP server tool call, returning a decision.
    function pre_mcp_tool_call_hook_interface(server_name, tool_name, tool_args, session_id) result(decision)
      character(len=*), intent(in) :: server_name, tool_name, tool_args, session_id
      character(len=:), allocatable :: decision
    end function pre_mcp_tool_call_hook_interface
  end interface

contains

  !> Set sensible defaults for client options.
  subroutine client_options_set_defaults(self)
    class(copilot_client_options), intent(inout) :: self

    if (.not. allocated(self%log_level)) then
      self%log_level = 'info'
    end if
    if (.not. allocated(self%cwd)) then
      self%cwd = '.'
    end if
    self%auto_start = .true.
    self%use_stdio  = .true.
  end subroutine client_options_set_defaults

end module copilot_types
